File: typescript/list/CircularLinkedList.ts

export class CircularLinkedListNode<T> {
    prev: CircularLinkedListNode<T> | null = null
    next: CircularLinkedListNode<T> | null = null
    constructor(public value: T) { }
}

export class CircularLinkedList<T> {
    head: CircularLinkedListNode<T> | null = null
    length: number = 0
    constructor() {

    }

    prepend(value: T): void {
        const newNode = new CircularLinkedListNode<T>(value);
        if (!this.head) {
            this.head = newNode;
            this.head.next = this.head;
            this.head.prev = this.head;
        } else {
            const tail = this.head.prev;
            newNode.next = this.head;
            newNode.prev = tail;
            tail!.next = newNode;
            this.head.prev = newNode;
            this.head = newNode

        }
        this.length++;
    }

    append(value: T): void {
        const newNode = new CircularLinkedListNode<T>(value);
        if (!this.head) {
            this.head = newNode;
            this.head.next = this.head;
            this.head.prev = this.head;
        } else {
            let tail = this.head.prev;
            tail!.next = newNode;
            newNode.prev = tail;
            newNode.next = this.head;
            this.head.prev = newNode;
        }
        this.length++;
    }

    insertAt(value: T, index: number): void {
        if (index < 0 || index > this.length) {
            throw new Error('Index out of bounds');
        }
        const newNode = new CircularLinkedListNode<T>(value);
        if (index === 0) {
            if (!this.head) {
                this.head = newNode;
                this.head.next = this.head;
                this.head.prev = this.head;
            }else {
                newNode.next = this.head;
                newNode.prev = this.head.prev;
                this.head.prev!.next = newNode;
                this.head.prev = newNode;
                this.head = newNode;
            }
        } else {
            let temp = this.head;
            for (let i = 0; i < index - 1; i++) {
                temp = temp!.next;
            }
            newNode.next = temp!.next;
            newNode.prev = temp;
            temp!.next = newNode;
            newNode.next!.prev = newNode;
        }
        this.length++;
    }

    deleteAt(index: number): void {
        if (index < 0 || index >= this.length) {
            throw new Error('Index out of bounds');
        }
        if (index === 0) {
            if (this.length === 1) {
                this.head = null;
            } else {
                const temp = this.head!;
                this.head = temp!.next;
                this.head!.prev = temp.prev;
                temp!.prev!.next = this.head;
            }
        } else {
            let temp = this.head;
            for (let i = 0; i < index - 1; i++) {
                temp = temp!.next;
            }
            temp!.next = temp!.next!.next;
            temp!.next!.prev = temp;
        }
        this.length--;
    }

    findAt(index: number): T | null {
        if (index < 0 || index >= this.length) {
            throw new Error('Index out of bounds');
        }
        let temp = this.head;
        for (let i = 0; i < index; i++) {
            temp = temp!.next;
        }
        return temp!.value
    }

    size() {
        return this.length
    }

    isEmpty() {
        return this.length === 0
    }

    isCircular() {
        let current = this.head
        while (current) {
            if (current.next === this.head && this.head?.prev === current) {
                return true
            }
            current = current.next
        }
        return false
    }

    *[Symbol.iterator]() {
        let current = this.head
        while (current) {
            yield current.value
            current = current.next
            if (current === this.head) {
                break
            }
        }
    }

    *reverseToArray() {

        let current = this.head?.prev
        while (current) {
            yield current.value
            current = current.prev
            if (current === this.head?.prev) {
                break
            }
        }
    }
}

File: Java/src/main/java/com/example/linkedlist/CircularLinkedList/CircularLinkedList.java
package com.example.linkedlist.CircularLinkedList;


public class CircularLinkedList {
    private CircularLinkedListNode head;

    private int length;

    public CircularLinkedList() {
        head = null;
        length = 0;
    }

    public boolean isEmpty() {
        return head == null;
    }

    public int size() {
        return length;
    }

    public boolean isCircular() {
        if (isEmpty())
            return false;
        boolean flag = false;
        CircularLinkedListNode temp = head;
        while (temp != null) {
            if (temp.next == head) {
                flag = true;
                break;
            }
            temp = temp.next;
        }
        return flag;
    }

    public void append(int data) {
        CircularLinkedListNode newNode = new CircularLinkedListNode(data);
        if (isEmpty()) {
            head = newNode;
            head.next = head;
            head.prev = head;
        } else {
            CircularLinkedListNode last = head.prev;
            last.next = newNode;
            newNode.prev = last;
            newNode.next = head;
            head.prev = newNode;
        }
        length++;
    }

    public void prepend(int data) {
        CircularLinkedListNode newNode = new CircularLinkedListNode(data);

        if (isEmpty()) {
            newNode.next = newNode;
            newNode.prev = newNode;
        } else {
            CircularLinkedListNode last = head.prev;
            head.prev = newNode;
            newNode.next = head;
            newNode.prev = last;
            last.next = newNode;
        }
        head = newNode;
        length++;
    }

    public void insertAt(int data, int index) {
        if (index < 0 || index > length) {
            throw new IndexOutOfBoundsException("index out if bounds");
        }

        CircularLinkedListNode newNode = new CircularLinkedListNode(data);

        if (index == 0) {
            if (isEmpty()) {
                newNode.next = newNode;
                newNode.prev = newNode;
            } else {
                newNode.next = head;
                newNode.prev = head.prev;
                head.prev.next = newNode;
                head.prev = newNode;
            }
            head = newNode;
        } else {
            CircularLinkedListNode temp = head;
            for (int i = 0; i < index - 1; i++) {
                temp = temp.next;
            }
            newNode.next = temp.next;
            newNode.prev = temp;
            temp.next.prev = newNode;
            temp.next = newNode;
        }
        length++;
    }

    public void deleteAt(int index) {
        if (index < 0 || index >= length) {
            throw new IndexOutOfBoundsException("index out if bounds");
        }

        if (index == 0) {
            if (length == 1) {
                head = null;
            } else {
                head.prev.next = head.next;
                head.next.prev = head.prev;
                head = head.next;
            }
        } else {
            CircularLinkedListNode temp = head;
            for (int i = 0; i < index - 1; i++) {
                temp = temp.next;
            }
            temp.next = temp.next.next;
            temp.next.prev = temp;
        }
        length--;
    }

    public int findAt(int index) {
        if (index < 0 || index >= length) {
            throw new IndexOutOfBoundsException("index out if bounds");
        }

        CircularLinkedListNode temp = head;
        for (int i = 0; i < index; i++) {
            temp = temp.next;
        }
        return (int) temp.data;
    }
    public int[] toArray() {
        if (isEmpty())
            return new int[0];

        int[] arr = new int[length];
        CircularLinkedListNode temp = head;
        for (int i = 0; i < length; i++) {
            arr[i] = (int) temp.data;
            temp = temp.next;
        }
        return arr;

    }

    public int[] reverseToArray() {
        if (isEmpty())
            return new int[0];

        int[] arr = new int[length];
        CircularLinkedListNode temp = head.prev;
        for (int i = 0; i < length; i++) {
            arr[i] = (int) temp.data;
            temp = temp.prev;
        }
        return arr;
    }
}

class CircularLinkedListNode {

    int data;
    CircularLinkedListNode next;
    CircularLinkedListNode prev;

    public CircularLinkedListNode(int data) {
        this.data = data;
    }
}
